### PR TITLE
Optimize HashInput functions to use scratch buffer provided by onflow/atree

### LINKED ISSUE #1258: Optimize `HashInput`
### Issue To Be Solved

`HashInput` can be optimized for increased performance and reduce heap allocation.

### Suggested Solution

Modify `HashInput` to use a scratch buffer provided to the function.

### PR DESCRIPTION
**Pull request by fxamacker**
Closes #1258 

## Description

Optimize `HashInput` functions to use a small scratch buffer provided by onflow/atree.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by turbolent — APPROVED** [review]
Nice! Thank you for optimizing this 👍

**Comment by turbolent** [comment]
@fxamacker It looks like coverage went down as the new cases where the input does not fit into the scratch are not covered by tests (e.g. https://app.codecov.io/gh/onflow/cadence/compare/1296/tree/runtime/interpreter/value.go#D1L352). 

Could you please add some tests with longer/larger values? Thank you!

**Comment by fxamacker** [comment]
> @fxamacker It looks like coverage went down as the new cases where the input does not fit into the scratch are not covered by tests (e.g. https://app.codecov.io/gh/onflow/cadence/compare/1296/tree/runtime/interpreter/value.go#D1L352).
> 
> Could you please add some tests with longer/larger values? Thank you!

Yikes!  Thanks for catching this! 👍  It would've been bad to merge with this mistake.  I added more tests.  Many thanks!

File: runtime/interpreter/value_test.go [hunk 1]
```diff
@@ -1116,82 +1118,242 @@ func TestGetHashInput(t *testing.T) {
 			value:    NewUIntValueFromUint64(10),
 			expected: []byte{byte(HashInputTypeUInt), 10},
 		},
+		"UInt min": {
```
Comment (turbolent) [inline runtime/interpreter/value_test.go:1121]: Great idea to use min & max values 👍

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**